[WIP] Catch and display path errors

kbrock opened the pull request:
Depends upon
- [x] #189

2.5 things:

- If a user specifies a bad path in a workflow, display a meaningful message.
- If a user specifies a path that fails a lookup in runtime, display a meaningful message.
- (This is not working with templates - not worse, but not informative/better either)

# Invalid path (validation)

```
    "Test1": {
      "Comment": "Parameter not a path",
      "Type":  "Pass",
      "InputPath": "nonparam",
      "End": true
    }
```

aws:

```
Field "InputPath" defined at "State Machine.Test1" is not a JSONPath
```

## Before

```
Path [nonparam] must start with "$"
```

## After

```
States.Test1 field "InputPath" value "nonparam" Path [nonparam] must start with "$"
```

# Path not found (runtime)

```
    "Test2": {
      "Type":  "Pass",
      "InputPath": "$.missing"
      "End": true
    }
```

aws:

```
Invalid path '$.missing' : No results for path: $['missing']
```

Before
======

```
{}
```

After
=====

```
{
  "Error":"States.RuntimeError",
  "Cause":"States.Test2 field \"InputPath\" value \"$.missing\" references an invalid value"
}
```

kbrock committed:
Better error messages for parsing bad input path, reference path, and payload templates

kbrock commented:
update:
- rubocop changes

update:
- removed some changes to process_output (code climate changes)

Fryguy commented:
This feels like adding a lot of "code" to just present error messages.  Is it possible to just not have the `wrap_runtime_error` as that feels like the method that is adding a lot of this overhead?

I am curious why the PathError itself can't contain the details within, since it is being raised from within the Path itself.  Additionally, is every instance of PathError just also an ExecutionError?  If so, could it just raise an ExecutionError instead, or, if not, can PathError just be a subclass of ExecutionError?

kbrock commented:
@Fryguy The problem is a generically thrown exception will not properly have the context for the error. So either the thing throwing the error needs more information, or we need to add information on the way up.

I'll add the state/path to the `Path`, so it can throw a better exception.

Fryguy commented:
I'm not sure I follow... It looks like the same thing is passed into path that's wrapped outside of it as well

kbrock commented:
WIP: I care more about Choice fixes than error display. incorporating these errors on top of #189

Fryguy commented on lib/floe.rb:
```diff
@@ -44,6 +44,17 @@ class Error < StandardError; end
+  class ExecutionError < Error
+    attr_reader :floe_error
+
+    def initialize(message, floe_error = "States.RuntimeError")
```
As Adam asked elsewhere, is States.RuntimeError an actual error?

agrare replied:
https://github.com/ManageIQ/floe/pull/189#pullrequestreview-2177893255 it isn't listed in the states-language spec but it is in the AWS StepFunctions docs and he fixed it to be `States.Runtime` to match the docs there

kbrock commented:
update:
- rebased onto latest choice payload validation

kbrock commented:
WIP: keeping up to date locally. waiting for choice to go through before putting too much effort into this. (they all conflict)

Also, my push more information into Path so it throws the fully qualified error instead of partial and having to manipulate the exception on the way up

kbrock committed:
WIP: think we want to move away from parsing at all inside workflow

also, team not thrilled with having the wrap stuff everywhere. so need to come back

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -170,7 +170,7 @@ def parse_field(field_name, data_type)
         # fetch a path at runtime
         def fetch_path(field_name, field_path, context, input)
-          value = field_path.value(context, input)
+          value =  wrap_runtime_error(field_name, field_path.to_s) { field_path.value(context, input) }
```
```suggestion
          value = wrap_runtime_error(field_name, field_path.to_s) { field_path.value(context, input) }
```

kbrock committed:
Better Runtime error detection for bad paths

kbrock commented:
update:
- rebased
- fixed rubocop

WIP: still WIP as I'd like to move `JSON.parse` out of the `initializer`s before handling the path errors - though can't remember right now why I felt that needed to be done first
